HPS/Heartland/Model/Payment.php

```diff
@@ -501,23 +908,144 @@ public function getToken(){
-        if (array_key_exists('_save_token_value',$data)){
+        if (array_key_exists('_save_token_value', $data)) {
             $this->_save_token_value = (int) $data['_save_token_value'];
         }
-        $this->log((int) $this->_save_token_value,'\HPS\Heartland\Model\Payment::saveMuToken ');
+        $this->log((int) $this->_save_token_value, '\HPS\Heartland\Model\Payment::saveMuToken ');
+
         return (int) $this->_save_token_value;
     }
 
+    /** returns additional_data element of paymentMethod
+     *
+     * @return array
+     */
+    private
+    function getAdditionalData()
+    {
+
+        static $data = [];
+        if (count($data) < 1) {
+            $data = (array) $this->getPaymentMethod();
+        }
+
+        return $this->elementFromArray($data, 'additional_data');
+    }
+
+    /** returns an element of associative array of data submitted via HTTP POST paymentMethod
+     *
+     * @return array
+     * */
+    private
+    function getPaymentMethod()
+    {
+        /**
+         * @var array $data
+         * Holds submited JSOn data in a PHP associative array
+         */
+        static $data = [];
+        if (count($data) < 1) {
+            $data = (array) HPS_Data::jsonData();
+        }
+        $this->log($data, 'HPS\Heartland\Model\Payment getPaymentMethod Method Called:  ');
+
+        return $this->elementFromArray($data, 'paymentMethod');
+    }
+
+    /** evaluates if an element exists and returns it
+     *
+     * @param $data
+     * @param $element
+     *
+     * @return array
+     */
+    private
+    function elementFromArray($data, $element)
+    {
+        $r = [];
+        if (key_exists($element, $data)) {
+            $r = (array) $data[ $element ];
+        }
+
+        return $r;
+    }
+
     /**
-     * validateSuToken performs regex based validation on the sungleUseToken
+     * Takes anything presented strips begining and ending whitespace and returns only string with no special characters
+     *
+     * @param $data
+     *
+     * @return string
+     */
+    private
+    function __sanitize($data)
+    {
+        return trim(filter_var(@print_r($data, true), FILTER_SANITIZE_STRING));
+    }
+
+    /**
+     * this method sets the instance  \HpsTokenData::$tokenValue
+     * If the \HPS\Heartland\Model\Payment::$_token_value that is sent is an integer only then we assume it is a
+     * primary key for hps_heartland_storedcard and perform a lookup
+     *
+     * @param \HpsTokenData $suToken
+     *
+     * @return \HpsTokenData
+     *
+     * @TODO: evaluate if something need to happen when no token is assigned. Probably safe to do nothing
+     */
+    private
+    function getToken(\HpsTokenData $suToken, $custID = null)
+    {
+        $this->log($this->_token_value, '\HPS\Heartland\Model\Payment::getToken Method initial value:  ');
+        $this->getTokenValue();
+        if (preg_match('/^[\w]{11,253}$/', (string) $this->_token_value) !== 1) {
```

Comment:
format checking removed and added the logic to get stored card if it's a numeric value